src/mpid/ch4/shm/xpmem/xpmem_seg.h

```diff
@@ -0,0 +1,644 @@
+                } else {
+                    direction = AVL_RIGHT;
+                }
+            }
+
+            mpi_errno = MPIDI_XPMEM_segtree_do_create_node(low, high, apid, seg_ptr);
+            if (mpi_errno != MPI_SUCCESS)
+                MPIR_ERR_POP(mpi_errno);
+
+            if (direction == AVL_LEFT)
+                sub_root->left = *seg_ptr;
+            else
+                sub_root->right = *seg_ptr;
+
+            (*seg_ptr)->parent = sub_root;
+            *voffset = 0;
+            tree->tree_size++;
+
+          stack_recovery:
+            mpi_errno = MPIDI_XPMEM_avl_do_update_node_info(sub_root);
+            if (mpi_errno != MPI_SUCCESS)
+                MPIR_ERR_POP(mpi_errno);
+
+            int lheight = sub_root->left == NULL ? 0 : sub_root->left->height;
+            int rheight = sub_root->right == NULL ? 0 : sub_root->right->height;
+            if (lheight - rheight > 1) {
+                /* left unbalanced */
+                MPIDI_XPMEM_seg_t *ct_node = sub_root->left;
+                int lct_height = ct_node->left == NULL ? 0 : ct_node->left->height;
+                if (lct_height + 1 == lheight) {
+                    /* Insert into left child of left child */
+                    /* Do right rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_do_right_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                } else {
+                    /* Insert into right child of left child */
+                    /* Do left-right rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_do_left_right_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                }
+            } else if (rheight - lheight > 1) {
+                /* right unbalanced */
+                MPIDI_XPMEM_seg_t *ct_node = sub_root->right;
+                int lct_height = ct_node->left == NULL ? 0 : ct_node->left->height;
+                if (lct_height + 1 == rheight) {
+                    /* Insert into left child of right child */
+                    /* Do right-left rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_do_right_left_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                } else {
+                    /* Insert into right child of right child */
+                    /* Do left rotation */
+                    mpi_errno = MPIDI_XPMEM_avl_do_left_rotation(sub_root, ct_node);
+                    if (mpi_errno != MPI_SUCCESS)
+                        MPIR_ERR_POP(mpi_errno);
+                }
+            }
+
+            if (!MPIDI_XPMEM_AVL_STACK_EMPTY(node_stack)) {
+                MPIDI_XPMEM_AVL_STACK_POP(node_stack, sub_root);
+                goto stack_recovery;
+            } else
+                break;
+
+        } while (1);
```

Comment:
I don't think we need any backup conditional since the program guarantees loop will exit in any case. 
The infinite loop here is used because I don't want to make a recursive call to search all the nodes in the AVL tree (inline won't allow and normal function call would be expensive). I also put the comments here to explain how it exits the loop. 

> /* This infinite loop will break in either two cases:
             * (1) when we search AVL tree and hit the segment cache, it means 
             * no insertion is needed, and we can jump to fn_exit and return cached
             * segment right away.
             * (2) when we cannot find cached segment, we need to attach and insert
             * a new segment into AVL tree; in this case, loop will exit when we 
             * finish insertion and AVL tree height adjustment.  */